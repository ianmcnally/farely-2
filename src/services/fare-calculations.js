import {
  TRANSACTION_MAX,
  RIDE_COST
} from '../constants'
import range from 'lodash.range'

const getPossibleFaresForSpendingRange = (currentBalance, maximumSpend) => {
  const ridesLeftOnCard = (currentBalance > RIDE_COST) ? Math.floor(currentBalance / RIDE_COST) : 0
  const amountOfRidesPossibleWithinBudget = Math.floor(maximumSpend / RIDE_COST)

  return range(ridesLeftOnCard + 1, amountOfRidesPossibleWithinBudget + 1).map(rideCount => ({
    rides: rideCount,
    fare: rideCount * RIDE_COST
  }))
}

const getCostsForFares = (currentBalance, fares) => fares.map(({ rides, fare }) => {
  const costOfFare = fare - currentBalance
  return { rides, cost: costOfFare }
})

const isDollarAmountAMultipleOfFive = dollarAmount => ((dollarAmount * 100) % 5) === 0

const isValidAmountToAdd = (value, maximumValue) => isDollarAmountAMultipleOfFive(value) && (value <= maximumValue)

export const costForFaresInBudget = (currentBalance = 0, maximumSpend = TRANSACTION_MAX) => {
  const maximumAmountToSpend = (maximumSpend > TRANSACTION_MAX) ? TRANSACTION_MAX : maximumSpend
  const possibleFares = getPossibleFaresForSpendingRange(currentBalance, maximumAmountToSpend)
  const possibleCosts = getCostsForFares(currentBalance, possibleFares)
  const validAmountsToAdd = possibleCosts.filter(({ cost }) => isValidAmountToAdd(cost, maximumAmountToSpend))
  return validAmountsToAdd
}
