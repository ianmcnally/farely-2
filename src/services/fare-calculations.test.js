import {
  BONUS_MIN,
  BONUS_PERCENT,
  TRANSACTION_MAX,
  RIDE_COST
} from '../constants'
import { costForFaresInBudget } from './fare-calculations'
import last from 'lodash.last'

describe('costForFaresInBudget', () => {

  describe('with a zero current balance', () => {
    const balance = 0
    const maximumSpend = 40
    let costs

    beforeAll(() => {
      costs = costForFaresInBudget(balance, maximumSpend)
    })

    it('returns costs from one ride to maximum spend', () => {
      expect(costs).toMatchSnapshot()
    })

  })

  describe('with a current balance below the maximum spend', () => {
    const balance = 25
    const maximumSpend = 40
    let costs

    beforeAll(() => {
      costs = costForFaresInBudget(balance, maximumSpend)
    })

    it('returns costs from the current balance', () => {
      expect(costs).toMatchSnapshot()
    })

  })

  describe('with a maximum spend above the transaction maximum', () => {
    const balance = 0
    const maximumSpend = 100
    let highestCost

    beforeAll(() => {
      const costs = costForFaresInBudget(balance, maximumSpend)
      highestCost = last(costs).cost
    })

    it('returns costs capped at the transaction max', () => {
      expect(highestCost).toBeLessThan(TRANSACTION_MAX)
    })

  })

})
