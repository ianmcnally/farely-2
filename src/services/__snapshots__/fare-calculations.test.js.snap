// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`costForFaresInBudget with a current balance below the maximum spend returns costs from the current balance 1`] = `
Array [
  Object {
    "cost": 2.5,
    "rides": 10,
  },
  Object {
    "cost": 5.25,
    "rides": 11,
  },
  Object {
    "cost": 8,
    "rides": 12,
  },
  Object {
    "cost": 10.75,
    "rides": 13,
  },
  Object {
    "cost": 13.5,
    "rides": 14,
  },
]
`;

exports[`costForFaresInBudget with a zero current balance returns costs from one ride to maximum spend 1`] = `
Array [
  Object {
    "cost": 2.75,
    "rides": 1,
  },
  Object {
    "cost": 5.5,
    "rides": 2,
  },
  Object {
    "cost": 8.25,
    "rides": 3,
  },
  Object {
    "cost": 11,
    "rides": 4,
  },
  Object {
    "cost": 13.75,
    "rides": 5,
  },
  Object {
    "cost": 16.5,
    "rides": 6,
  },
  Object {
    "cost": 19.25,
    "rides": 7,
  },
  Object {
    "cost": 22,
    "rides": 8,
  },
  Object {
    "cost": 24.75,
    "rides": 9,
  },
  Object {
    "cost": 27.5,
    "rides": 10,
  },
  Object {
    "cost": 30.25,
    "rides": 11,
  },
  Object {
    "cost": 33,
    "rides": 12,
  },
  Object {
    "cost": 35.75,
    "rides": 13,
  },
  Object {
    "cost": 38.5,
    "rides": 14,
  },
]
`;
